### PR TITLE
Add partial support for converting OData to SQLite

### PR DESCRIPTION
**Pull request by shrishrirang**
- Not all queries will generate sqlite yet. More work needs to be done before we get there.
- SQLFormatter's format method now returns an array of sql statements. The statements can be merged into a single statement according to the sql flavor being used.

### CODE REVIEW

File: src/data/sql/execute.js [hunk 1]
```diff
@@ -24,21 +24,39 @@ module.exports = function (config, statement) {
...
-                    request.input(parameter.name, parameter.value);
-            });
+        var statements = sql;
+        if (sql.constructor !== Array) {
+            statements = [sql];
         }
+        
+        // We expect multiple results if there are multiple statements to be executed or if
+        // 'multiple' results are explicitly requested
+        request.multiple = statements.length > 1 ||  statements[0].multiple;
```
Comment (danderson00) [inline src/data/sql/execute.js]: This is a bit of a hack (checking the first statement for "multiple"), I'm not really comfortable with it. We need to come up with a better solution.

File: src/data/sql/query/format.js [hunk 2]
```diff
@@ -582,3 +595,9 @@ function isConstantOfType(expr, type) {
 function systemPropertyToColumnName(propertyName) {
     return '__' + propertyName;
 }
+
+function Statement() {
```
Comment (danderson00) [inline src/data/sql/query/format.js]: Don't use a "type" here. It's entirely unnecessary.

File: src/data/sql/query/format.js [hunk 3]
```diff
@@ -42,51 +37,74 @@ var SqlFormatter = types.deriveClass(ExpressionVisitor, ctor, {
...
 
-        if (query.skip >= 0 && query.take >= 0 && query.skip !== null && query.take !== null) {
-            this.sql = this._formatPagedQuery(query);
-        } else {
-            this.sql = this._formatQuery(query);
-        }
+        var statements = [];
+        this.paramNumber = 0;
+        
+        this.statement = new Statement();
```
Comment (danderson00) [inline src/data/sql/query/format.js]: Don't use a "type" here

File: src/data/sql/execute.js (hunk 1 above)
Comment (danderson00) [inline src/data/sql/execute.js]: In fact, these changes won't cut it when we add support for the SQLite provider - we can't merge statements into a single query when using SQLite. However, I'm happy to continue with this approach for now to unblock your cordova work. If we change this module down the track, it shouldn't impact the work you are doing.

Comment (shrishrirang) [inline src/data/sql/execute.js]: Yes, changing how execute.js consumes the output of the SQLFormatter will not impact other clients like cordova sqlite store.

File: src/data/sql/query/format.js (hunk 3 above)
Comment (shrishrirang) [inline src/data/sql/query/format.js]: Changed to a factory function statement()

File: src/data/sql/query/format.js (hunk 2 above)
Comment (shrishrirang) [inline src/data/sql/query/format.js]: changed to a factory function statement()

File: src/data/sql/query/format.js [hunk 4]
```diff
@@ -42,51 +37,74 @@ var SqlFormatter = types.deriveClass(ExpressionVisitor, ctor, {
...
 
-        if (query.skip >= 0 && query.take >= 0 && query.skip !== null && query.take !== null) {
-            this.sql = this._formatPagedQuery(query);
-        } else {
-            this.sql = this._formatQuery(query);
-        }
+        var statements = [];
+        this.paramNumber = 0;
+        
+        this.statement = statement();
```
Comment (danderson00) [inline src/data/sql/query/format.js]: this.statement = { multiple: true, parameters: [] }
    if(/* includeTotalCount logic below */)
      this.statement.sql = 'blah';
    else
      this.statement.sql = 'other blah';
    
    statements.push(this.statement)

No factory function required at all.

Comment (shrishrirang) [inline src/data/sql/query/format.js]: There is no if-else. We are creating 2 statement instances (conditionally). I can initialize the way you mentioned twice but that's creating scope for errors in the future - someone can update one initialization but miss the other.

Comment (danderson00) [inline src/data/sql/query/format.js]: Fair enough. Still, having a constructor or factory function to initialise three properties to blank values is overkill. Just do something like:

statements.push({ sql: 'blah' })
if(includeTotalCount)
  statements.push({ sql: this._formatCountQuery() })

Comment (shrishrirang) [inline src/data/sql/query/format.js]: The _format* functions need the statement object for parameter generation (_createParameter method adds parameters to this.statement).

Comment (danderson00) [inline src/data/sql/query/format.js]: I know, this was just example code. This is unrelated to whether or not there is a factory or constructor. The main drive behind these comments is to remind you that you're an author - try to put yourself in the perspective of someone who has never read the code before - the file is around 600 lines long - having to scroll to the end of the file and back up just to understand that this function is creating a simple object with blank values makes it harder to understand. Always write with the author mentality (yep, there are plenty of examples where I have failed at this haha)

Comment (shrishrirang) [inline src/data/sql/query/format.js]: done

Comment (danderson00) [inline src/data/sql/query/format.js]: This is how the code looks now:

    this.statement = { sql: '', parameters: [], multiple: true };
    this.statement.sql = this._formatQuery(query).trim();
    statements.push(this.statement);

    if (query.inlineCount === 'allpages' || query.includeTotalCount) {
        this.statement = { sql: '', parameters: [], multiple: true };
        this.statement.sql = this._formatCountQuery(helpers.formatTableName(this.schemaName, query.table), query).trim();
        statements.push(this.statement);
    }

Would you agree that the following is clearer?

    statements.push({ 
        sql: this._formatQuery(query).trim(),
        parameters: [],
        multiple: true
    });

    if (query.inlineCount === 'allpages' || query.includeTotalCount)
        statements.push({
            sql: this._formatCountQuery(helpers.formatTableName(this.schemaName, query.table), query).trim(),
            parameters: [],
            multiple: true
        });

Also, the fact it is assigning a statement to this.statement more than once is really confusing.

**Comment by shrishrirang** [comment]
Ping.

**Pull request closed without merging**